### ISSUE TITLE
feat(arch): add architecture + GGUF support for cross-encoder + LLM-native rerankers (justrank)

### ISSUE THREAD

**Issue by sweengineeringlabs** [sweengineeringlabs/rust-deeplearning#4]
## Summary

Stand up a `justrank` service (sibling to `justembed`) that exposes
`POST /v1/rerank` for retrieval-pipeline reranking. Two backend
families need to land in `swe-llmmodel-arch-*` to make this possible.
Filing this here so the ML team owns the architecture work; the
service crate and the rag-side wiring (swerag → llmboot) are out of
scope for this issue and tracked separately.

## Motivation

llmboot's RAG pipeline (via swerag) currently exposes
`RetrieveStrategy::{Hybrid, Semantic, Keyword}`. Hybrid is implemented
end-to-end (vector + BM25 + RRF) but there is no reranker pass after
fusion. Adding a cross-encoder or LLM-native rerank step on the top-N
RRF candidates is the standard accuracy lever — typical gains are
measurable, particularly for paraphrase-heavy queries where pure
embeddings underperform.

## Scope split

The work splits cleanly into **architecture readiness** (this repo's
problem, this issue) and **service/wiring** (justrank repo + swerag +
llmboot, separate issues).

## What this issue tracks — architecture readiness

Two reranker families. v1 ships the encoder path, v2 follows.

### v1 — encoder cross-encoder (priority)

**Target model**: `BAAI/bge-reranker-v2-m3` — ~568M params,
multilingual XLM-RoBERTa fine-tuned for relevance scoring,
Apache-2.0, GGUF quantizations exist on HF. Current SOTA among
small open-weight cross-encoders.

Architecturally an encoder + a classifier head:

```
[CLS] query [SEP] document [SEP]
   |
   v
XLM-RoBERTa encoder        <- already mostly covered by arch-bert /
                              arch-nomic-bert; verify XLM-R variant
                              handled (RoBERTa positional encoding,
                              SentencePiece tokenizer)
   |
   v
CLS hidden state           <- PoolingStrategy::Cls (verify existence
                              in swe-llmmodel-layers; add if missing)
   |
   v
Linear(hidden_dim -> 1)    <- NEW: classifier head, not present in
                              embedding-only models. The GGUF must
                              carry these weights and the model
                              loader must wire them
   |
   v
relevance score (scalar)
```

**Required work**:

1. **`arch-bert` (or new `arch-xlm-roberta`)** — confirm or extend so
   it handles XLM-RoBERTa's tokenizer + positional embedding.
   bge-reranker-v2-m3 is XLM-RoBERTa-derived, not vanilla BERT.
2. **Classifier head support** — extend `LanguageModel` (or a sibling
   trait like `SequenceClassifier`) with a method that runs the
   encoder + reads the CLS hidden state + projects through a linear
   head, returning a single score per input pair.
3. **GGUF coverage** — verify `swe-llmmodel-gguf` parses
   bge-reranker-v2-m3's quantized variants (Q4_K_M, Q8_0, F16) AND
   surfaces the classifier head tensor. Some HF -> GGUF conversion
   scripts strip the head — the loader must fail loudly if the head
   is absent rather than silently producing wrong scores.
4. **`PoolingStrategy::Cls`** — confirm exists in
   `swe-llmmodel-layers`; add if not. Different strategy from `Mean`.

### v2 — decoder LLM-native rerankers (follow-up)

**Target models**: RankLLaMA, RankMistral, Qwen3-Reranker
(0.6B / 4B / 8B variants). These are decoder-only LLMs fine-tuned
for relevance scoring via yes/no logit extraction:

```
prompt: "Given query: {q} document: {d}. Is this relevant? Answer:"
   |
   v
LanguageModel::forward(input_ids) -> logits [batch, seq, vocab]
   |
   v
last-position logits [batch, vocab]
   |
   v
softmax -> probs[yes_token_id] = relevance score
```

The `LanguageModel` trait already exposes `forward()` so the
interface fits. The gap is **arch coverage**:

| Reranker | Base | Status |
|---|---|---|
| RankLLaMA | LLaMA 7B/13B | yes — `arch-llama` exists; verify with parity test |
| RankMistral | Mistral 7B | maybe — sliding-window attention; depends on whether `arch-llama` covers it or needs a new `arch-mistral` |
| Qwen3-Reranker | Qwen3 0.6B/4B/8B | no — needs new `arch-qwen3` (Qwen3 has its own RoPE config + GQA layout) |

v2 work, in priority order:
1. Verify `arch-llama` runs RankLLaMA correctly (parity test against a reference inference).
2. Decide: extend `arch-llama` to cover Mistral, or add `arch-mistral`.
3. Add `arch-qwen3`. This is the most useful long-term target given
   model size (0.6B variant runs on CPU at acceptable latency) and
   is the SOTA LLM-native reranker.

### v1

- `arch-bert` (or `arch-xlm-roberta`) loads `bge-reranker-v2-m3`
  GGUFs (Q4_K_M minimum) and produces forward-pass output that
  matches a reference (HF transformers or sentence-transformers)
  within an agreed tolerance.
- Classifier head loads from GGUF and is callable via the model
  surface. Loader fails explicitly if the head tensor is absent.
- `PoolingStrategy::Cls` exists and selects the CLS hidden state.
- Parity test: 50 (query, doc) pairs; relevance scores from this
  stack vs reference implementation match within tolerance.

### v2

- `arch-llama` parity test on a RankLLaMA GGUF (50 prompts).
- Decision recorded on Mistral handling (extend vs new arch).
- `arch-qwen3` lands with parity test on Qwen3-Reranker-0.6B.

## Out of scope (other issues / repos)

- The **justrank service crate** itself (workspace clone of justembed
  with `POST /v1/rerank` router, classifier-head wiring, tokenizer
  helpers for the `[CLS] q [SEP] d [SEP]` concatenation) — separate
  issue once architectures land.
- **swerag** changes (`SweRerankClient`, plumbing through
  `core/retrieve.rs` as a post-RRF pass with a configurable
  `rerank_multiplier` for top-N selection) — issue in the rag repo.
- **llmboot** changes (`rerank: Option<bool>` on `RagAgentConfig`,
  pass-through on `RetrieveRequest`) — issue in llmboot, blocked
  on swerag.

## References

- bge-reranker-v2-m3: https://huggingface.co/BAAI/bge-reranker-v2-m3
- Qwen3-Reranker: https://huggingface.co/Qwen/Qwen3-Reranker-0.6B
- RankLLaMA: https://huggingface.co/castorini/rankllama-v1-7b-lora-passage
- justembed (architectural template): see this repo's
  `justembed/{conversion,systemd}/`
